control-plane/agents/src/bin/core/controller/scheduling/mod.rs

```diff
@@ -97,6 +99,61 @@ pub(crate) trait ResourceFilter: Sized {
     }
 }
 
+/// Represents a sort criteria to be passed to a sort builder.
+pub(crate) struct SortCriteria {
+    criteria: Criteria,
+    grading: ValueGrading,
+    value_fn: Box<dyn Fn(&PoolItem) -> Value>,
+}
+
+impl SortCriteria {
+    /// Create a new sort criteria.
+    pub(crate) fn new(
+        criteria: Criteria,
+        grading: ValueGrading,
+        value_fn: impl Fn(&PoolItem) -> Value + 'static,
+    ) -> Self {
+        SortCriteria {
+            criteria,
+            grading,
+            value_fn: Box::new(value_fn),
+        }
+    }
+}
+
+/// Builds a weighted sorting comparator, with the various sort criterias being added to it.
+pub(crate) struct SortBuilder {
+    sort_criterias: Vec<SortCriteria>,
+}
+
+impl SortBuilder {
+    /// Create a new sort builder.
+    pub(crate) fn new() -> Self {
+        SortBuilder {
+            sort_criterias: Vec::new(),
+        }
+    }
+
+    /// Add sort criteria to the builder.
+    pub(crate) fn with_sort_criteria(mut self, sort_criteria: fn() -> SortCriteria) -> Self {
```

Comment:
```suggestion
    pub(crate) fn with_criteria(mut self, sort_criteria: fn() -> SortCriteria) -> Self {
```
We're already the `Sort`Builder